### PR TITLE
[TableGen] Refactor Intrinsic handling in TableGen

### PR DESCRIPTION
**Pull request by jurahul**
CodeGenIntrinsic changes:
  - Use `const` Record pointers, and `StringRef` when possible.
  - Default initialize several fields with their definition instead of in the constructor.
  - Simplify various string checks in the constructor using StringRef starts_with()/ends_with() functions.
  - Eliminate first argument to `setDefaultProperties` and use `TheDef` class member instead.

IntrinsicEmitter changes:
  - Emit `namespace llvm::Intrinsic` instead of nested namespaces.
  - End generated comments with a .
  - Use range based for loops, and early continue within loops.
  - Emit `static constexpr` instead of `static const` for arrays.
  - Change `compareFnAttributes` to use std::tie() to compare intrinsic attributes and return a default value when all attributes are equal.

STLExtras:
  - Add std::replace wrapper which takes a range.

### CODE REVIEW

File: llvm/utils/TableGen/IntrinsicEmitter.cpp
Comment (jurahul) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:603]: I had clang test failures due to an earlier version of this change that used DenseMap. I added this comment to explain why both the maps need to be ordered.

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [inline polly/lib/Support/GICHelper.cpp:147]: For some reason, the compiler is resolving this replace() to the one I added and not the static one defined above. Hence this change. Also, using std::replace to replace one char with another is likely faster than using string based replacement.

**Comment by jurahul** [comment]
All checks are clean now, so this is ready for review.

File: llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp
Comment (arsenm) [inline llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp:92]: You can use combine with above with consume_front

Comment (jurahul) [reply llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp:92]: Thanks, I did not know about `consume_front`. Let me try to incorporate it.

File: polly/lib/Support/GICHelper.cpp
Comment (kuhar) [reply polly/lib/Support/GICHelper.cpp:147]: The reason is probably that the one in stlextras doesn't require type conversions while the static one above needs to convert to StringRef

File: llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp
Comment (jurahul) [reply llvm/utils/TableGen/Basic/CodeGenIntrinsics.cpp:92]: Done.

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp:147]: I see, so if I did an explicit StringRef("") it might resolve correctly. But I feel it might be confusing to a casual reader, so may be the name change is not bad. And also for the single char ones, we get to use the llvm::replace.

**Review by arsenm — APPROVED** [review]

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp:147]: This works as well:

```
  llvm::replace(str, '.', '_');
  llvm::replace(str, '\"', '_');
  replace(str, StringRef(" "), StringRef("__"));
  replace(str, StringRef("=>"), StringRef("TO"));
  llvm::replace(str, '+', '_');

```

Happy to incorporate it if that is preferable.

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp:137]: Nope, see comment above, this replace now conflicts with the one added in STLExtras.

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp:147]: @arsenm FYI.

Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp:147]: @kuhar please LMK if we want to do this change.

**Comment by jurahul** [comment]
What's the convention in terms of approvals needed? Do we need atleast 2 approvals or something before merging the change? In any case there is an unresolved thread above, so will wait, but wanted to know.

File: llvm/utils/TableGen/Basic/CodeGenIntrinsics.h
Comment (topperc) [inline llvm/utils/TableGen/Basic/CodeGenIntrinsics.h:146]: Since you're here. Can you change the semicolon at the end of this comment to a period.

Comment (jurahul) [reply llvm/utils/TableGen/Basic/CodeGenIntrinsics.h:146]: Will do.

Comment (jurahul) [reply llvm/utils/TableGen/Basic/CodeGenIntrinsics.h]: Done.

File: polly/lib/Support/GICHelper.cpp
Comment (kuhar) [reply polly/lib/Support/GICHelper.cpp:147]: I have any stakes here to be honest, just lurking here :)

Comment (kuhar) [reply polly/lib/Support/GICHelper.cpp:147]: The implementation of `llvm::replace` looks good to me

Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp]: Updated to use that version. Resolving.

File: polly/lib/Support/GICHelper.cpp
Comment (jurahul) [reply polly/lib/Support/GICHelper.cpp]: Resolving as well.

**Comment by jayfoad** [comment]
> What's the convention in terms of approvals needed? Do we need atleast 2 approvals or something before merging the change? In any case there is an unresolved thread above, so will wait, but wanted to know.

See https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted but to summarize, one approval is enough so long as there are no outstanding concerns from other reviewers.

**Comment by jayfoad** [comment]
Thanks for doing this. I am strongly in favor of modernizing the coding style in utils/TableGen.

**Comment by jurahul** [comment]
Thanks, I think I have addressed all review comments, so will go ahead and merge.

**Pull request merged**
